src/rulesets/LinearAlgebra/structured.jl

```diff
@@ -47,6 +47,225 @@ end
+            df = λi -> frule((Zero(), One()), $func, λi)
+            λ, U = eigen(A)
+            fλ_df_dλ = (df).(λ)
+            fλ = first.(fλ_df_dλ)
+            Y = _symherm!($func, U * Diagonal(fλ) * U', :U)
+            ∂Y = Thunk() do
+                df_dλ = last.(unthunk.(fλ_df_dλ))
+                ∂Λ = U' * ΔA * U
+                U′∂YU = _muldiffquotmat($func, λ, fλ, df_dλ, ∂Λ)
+                return _symherm!(Y, U * U′∂YU * U')
+            end
+            return Y, ∂Y
+        end
+
+        function rrule(::typeof($func), A::LinearAlgebra.RealHermSymComplexHerm)
+            df = λi -> frule((Zero(), One()), $func, λi)
+            λ, U = eigen(A)
+            fλ_df_dλ = (df).(λ)
+            fλ = first.(fλ_df_dλ)
+            Y = _symherm!($func, U * Diagonal(fλ) * U', :U)
+            function $(Symbol("$(func)_pullback"))(ΔY)
+                ∂A = Thunk() do
+                    df_dλ = unthunk.(last.(fλ_df_dλ))
+                    ∂fΛ = U' * _realifydiag(ΔY) * U
+                    U′∂AU = _muldiffquotmat($func, λ, fλ, df_dλ, ∂fΛ)
+                    return _symherm(A, U * U′∂AU * U')
+                end
+                return NO_FIELDS, ∂A
+            end
+            return Y, $(Symbol("$(func)_pullback"))
+        end
+    end
+end
+
+function frule((_, ΔA), ::typeof(sincos), A::LinearAlgebra.RealHermSymComplexHerm)
+    λ, U = eigen(A)
+    sinλ, cosλ = sin.(λ), cos.(λ)
+    sinA = _symherm!(A, U * Diagonal(sinλ) * U', :U)
+    cosA = _symherm!(A, U * Diagonal(cosλ) * U', :U)
+    sincosA = (sinA, cosA)
+    ∂sincosA = Thunk() do
+        ∂Λ = U' * ΔA * U
+        U′∂sinAU = _muldiffquotmat(sin, λ, sinλ, cosλ, ∂Λ)
+        ∂sinA = _symherm!(sinA, U * U′∂sinAU * U')
+        U′∂cosAU = _muldiffquotmat(cos, λ, cosλ, -sinλ, ∂Λ)
+        ∂cosA = _symherm!(cosA, U * U′∂cosAU * U')
+        return Composite{typeof(sincosA)}(∂sinA, ∂cosA)
+    end
+    return sincosA, ∂sincosA
+end
+
+function rrule(::typeof(sincos), A::LinearAlgebra.RealHermSymComplexHerm)
+    λ, U = eigen(A)
+    sinλ, cosλ = sin.(λ), cos.(λ)
+    sinA = _symherm!(A, U * Diagonal(sinλ) * U', :U)
+    cosA = _symherm!(A, U * Diagonal(cosλ) * U', :U)
+    sincosA = (sinA, cosA)
+    function sincos_pullback(ΔsincosA)
+        ∂A = Thunk() do
+            ΔsinA, ΔcosA = ΔsincosA
+            ∂sinΛ, ∂cosΛ = U' * _realifydiag(ΔsinA) * U, U' * _realifydiag(ΔcosA) * U
+            inds = eachindex(λ)
+            U′∂AU = @inbounds begin
+                _diffquot.(sin, inds, inds', Ref(λ), Ref(sinλ), Ref(cosλ)) .* ∂sinΛ .+
+                _diffquot.(cos, inds, inds', Ref(λ), Ref(cosλ), Ref(-sinλ)) .* ∂cosΛ
+            end
+            return _symherm(A, U * U′∂AU * U')
+        end
+        return NO_FIELDS, ∂A
+    end
+    return sincosA, sincos_pullback
+end
+
+# difference quotient, i.e. Pᵢⱼ = (f(λᵢ) - f(λⱼ)) / (λᵢ - λⱼ), with f'(λᵢ) when i==j
+Base.@propagate_inbounds function _diffquot(f, i, j, λ, fλ, df_dλ)
```

Comment:
is `f` needed here (since we get `fλ` and `df_dλ`)?

Reply:
No, not anymore.